kolibri/logger/tasks.py

```diff
@@ -0,0 +1,87 @@
+import logging as logger
+import threading
+import time
+
+from django.db import transaction
+
+logging = logger.getLogger(__name__)
+
+
+class AsyncLogQueue():
+
+    def __init__(self):
+
+        # Value in seconds to determine the sleep time between log saving batches
+        self.log_saving_interval = 5
+
+        # Where new log saving functions are appended
+        self.queue = []
+
+        # Where the to be executed log saving functions are stored
+        # once a batch save has been invoked
+        self.running = []
+
+    def append(self, fn):
+        """
+        Convenience method to append log saving function to the current queue
+        """
+        self.queue.append(fn)
+
+    def toggle_queue(self):
+        """
+        Method to swap the queue and running, to allow new log saving functions
+        to be added to the queue while previously added functions are being executed
+        and cleared without fear of race conditions dropping saves.
+        """
+        old_queue = self.queue
+        new_queue = self.running
+        self.queue = new_queue
+        self.running = old_queue
+
+    def clear_running(self):
+        """
+        Reset the running list to drop references to already executed log saving functions
+        """
+        self.running = []
+
+    def run(self):
+        """
+        Execute any log saving functions in the self.running list
+        """
+        if self.running:
+            # Do this conditionally to avoid opening an unnecessary transaction
+            with transaction.atomic():
+                for fn in self.running:
+                    try:
+                        fn()
+                    except Exception as e:
+                        # Catch all exceptions and log, otherwise the background process will end
+                        # and no more logs will be saved!
+                        logging.debug("Exception raised during background log saving: ", e)
```

Comment:
Should this be a `logging.warning`? Seems pretty severe when our logs fail to save.

Reply:
Fair!